File: RB101_RB109_small_problems/medium_1/rotation_p2.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: two integers (number num, number of last digits n to rotate in num)
-  Output: num with the appropriate number of digits rotated

---

**Problem Domain:**
- A single rotation is shifting the leading digit to the last position in the section to be rotated.
---

**Implicit Requirements:**
- Rotating a single digit will result in the original number num to be returned.
- May use rotate_array from the previous exercise.
- Assume n is always a positive integer.
---

**Clarifying Questions:**

1. Will the first number argument ever be negative or a float?
  - Do not consider negative or floats as possible inputs.

---

**Mental Model:**

Convert number num to a string.
Determine section of number to be rotated (last n spots in num_string)
Split num_string into separate strings - first string will be numbers TO NOT rotate, second string will be numbers TO rotate.
Rotate numbers.
Concatenate together unchanged first string with rotated second string.
Convert concatenation to integer.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

rotate_rightmost_digits(735291, 1) == 735291
rotate_rightmost_digits(735291, 2) == 735219
rotate_rightmost_digits(735291, 3) == 735912
rotate_rightmost_digits(735291, 4) == 732915
rotate_rightmost_digits(735291, 5) == 752913
rotate_rightmost_digits(735291, 6) == 352917

---

Data Structure
--------------

- Use strings for initial manipulation & splitting apart which section to rotate vs. not rotate.
- Use arrays to perform rotation.


---

Algorithm
---------

- Convert number num to a string.
- Determine section of number to be rotated (last n spots in num_string)
- Split num_string into separate strings - first string will be numbers TO NOT rotate, second string will be numbers TO rotate.
- Rotate numbers of second string.
- Concatenate together unchanged first string with rotated second string.
- Convert concatenation to integer.

Code
----

=end

def rotate_array(input_arr)
  return [] if input_arr.empty?
  rotated_array = input_arr.clone
  rotated_array << rotated_array.shift
end

def rotate_string(string)
  rotate_array(string.split('')).join
end

def rotate_rightmost_digits(num, n)
  num_string = num.to_s
  non_rotated_section = num_string[0...-n]
  section_to_rotate = num_string[-n..]

  rotated_section = rotate_string(section_to_rotate)

  (non_rotated_section + rotated_section).to_i
end

p rotate_rightmost_digits(735291, 1) == 735291
p rotate_rightmost_digits(735291, 2) == 735219
p rotate_rightmost_digits(735291, 3) == 735912
p rotate_rightmost_digits(735291, 4) == 732915
p rotate_rightmost_digits(735291, 5) == 752913
p rotate_rightmost_digits(735291, 6) == 352917

#LS Solution
def rotate_rightmost_digits(number, n)
  all_digits = number.to_s.chars
  all_digits[-n..-1] = rotate_array(all_digits[-n..-1])
  all_digits.join.to_i
end

File: JS210_small_problems/medium_1/rotation_p2.js
"use strict";
/*

Write a function that rotates the last n digits of a number. For the rotation, rotate by one digit to the left, moving the first digit to the end.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: Two numbers. The first number is a `number` while the second number is the number of digits `n` to rotate left by one spot in the first argument `number`.
-  Output: Return a number with the last `n` digits in `number` rotated to the left.

---

**Problem Domain:**

Nothing specific.

---

**Explicit Requirements:**
- Return a number.
- Take two numbers as inputs. First is `number` seconds is `n`.
- When rotating the last `n` digits in `number`, rotate to the left. The leftmost digit in the last `n` digits of `number` gets placed at the end.
---


**Implicit Requirements:**
- Function must be named rotateRightmostDigits
---

**Clarifying Questions:**

1. Will `n` always be a positive integer greater than or equal to 0?
  Yes.
2. Will `n` always be less than or equal to the total number of digits in `number`? i.e. rotateRightmostDigits(123, 4) is not valid?
  `n` can be greater than the number of digits in `number`. But in these cases, act as if `n` is just the number of digits in `number`.
3. Will `n` and `number` always be numbers? i.e. do I need to validate input?
  Assume `n` and `number` will always be numbers.

---

**Mental Model:**

- Validate input (check if `n` is 0 or greater than the number of digits in `number`)
- Split `number` into an array of digits.
- Rotate only the last `n` digits in the array of digits.
- Rejoin the digits together and return the result as a number.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

rotateRightmostDigits(735291, 0);      // 735291
rotateRightmostDigits(735291, 1);      // 735291
rotateRightmostDigits(735291, 2);      // 735219
rotateRightmostDigits(735291, 3);      // 735912
rotateRightmostDigits(735291, 4);      // 732915
rotateRightmostDigits(735291, 5);      // 752913
rotateRightmostDigits(735291, 6);      // 352917
rotateRightmostDigits(735291, 7);      // 352917

rotateRightmostDigits(43, 0);         // 43
rotateRightmostDigits(43, 1);         // 43
rotateRightmostDigits(43, 2);         // 34
rotateRightmostDigits(43, 3);         // 34


rotateRightmostDigits(400, 2);        // 400
rotateRigthmostDigits(400, 3);        // 4
---

Data Structure
--------------

Use strings to contain each digit.

"4"
"3"

Use an array to contain the strings of individual digits. Will allow use of list processing methods.

["4", "3"]

---

Algorithm
---------

- Validate input (check if `n` is 0 or 1 or greater than the number of digits in `number`)
  - If `n` is less than or equal to 1, return the input number.
  - If `n` is greater than the number of digits in `number`
    - Reassign `n` to the number of digits in `number`.
- Split `number` into an array of digits.
  - Convert `number` to a string. Split the string into an array of strings containing a single digit.
- Rotate only the last `n` digits in the array of digits.
  - Determine the array index that contains the leftmost digit in the `n` last digits. Assign to variable `leftmostDigitIndex`
    - leftmostDigitIndex = array length - `n`
  - Destructively splice the digit at the `leftmostDigitIndex`. Push it onto the end of the array of digits. 
- Rejoin the digits together and return the result as a number.
  - Join the digits together. Convert the resulting string to a number. Return the number.

Code
----
*/

function rotateRightmostDigits(number, n) {
  if (n <= 1) {
    console.log(number);
    return number;
  } else if (n > String(number).length) {
    n = String(number).length;
  }

  let digits = String(number).split('');

  let leftmostDigitIndex = digits.length - n;

  digits = digits.concat(digits.splice(leftmostDigitIndex, 1));

  let result = Number(digits.join(''));
  console.log(result);
  return result;
}

rotateRightmostDigits(735291, 0);      // 735291
rotateRightmostDigits(735291, 1);      // 735291
rotateRightmostDigits(735291, 2);      // 735219
rotateRightmostDigits(735291, 3);      // 735912
rotateRightmostDigits(735291, 4);      // 732915
rotateRightmostDigits(735291, 5);      // 752913
rotateRightmostDigits(735291, 6);      // 352917
rotateRightmostDigits(735291, 7);      // 352917

rotateRightmostDigits(43, 0);         // 43
rotateRightmostDigits(43, 1);         // 43
rotateRightmostDigits(43, 2);         // 34
rotateRightmostDigits(43, 3);         // 34


rotateRightmostDigits(400, 2);        // 400
rotateRightmostDigits(400, 3);        // 4